In [1]:
##first import the packages needed for the program

import pandas as pd             #best packages for array calculation    
import numpy as np              #numpy is a package for mathmatical calculation of dataframe
from sklearn import preprocessing      #this package is for scaling and process data
from sklearn import cross_validation   #training and testing data crossvalidation is 
                                       #basically applied for known data to find accuracy
from sklearn import neighbors          #this is the KNN packages 

from sklearn.ensemble import GradientBoostingClassifier  ## gradient boosting packages
                                                         #i use it from experience ;
                                                         #some case it works very good
from sklearn.ensemble import RandomForestClassifier      #random forest packages 

from sklearn import svm


/root/anaconda2/lib/python2.7/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [2]:
#first the dataframe will be created
df=pd.read_csv('breast-cancer-wisconsin.csv')

In [3]:
##data frame is created we can see the peak of the datasets
df

id  clump_thickness  unif_cell_size  unif_cell_shape  marg_adhesion  \
0    1000025                5               1                1              1   
1    1002945                5               4                4              5   
2    1015425                3               1                1              1   
3    1016277                6               8                8              1   
4    1017023                4               1                1              3   
5    1017122                8              10               10              8   
6    1018099                1               1                1              1   
7    1018561                2               1                2              1   
8    1033078                2               1                1              1   
9    1033078                4               2                1              1   
10   1035283                1               1                1              1   
11   1036172                2               1                1              1   
12   1041801                5               3                3              3   
13   1043999                1               1                1              1   
14   1044572                8               7                5             10   
15   1047630                7               4                6              4   
16   1048672                4               1                1              1   
17   1049815                4               1                1              1   
18   1050670               10               7                7              6   
19   1050718                6               1                1              1   
20   1054590                7               3                2             10   
21   1054593               10               5                5              3   
22   1056784                3               1                1              1   
23   1057013                8               4                5              1   
24   1059552                1               1                1              1   
25   1065726                5               2                3              4   
26   1066373                3               2                1              1   
27   1066979                5               1                1              1   
28   1067444                2               1                1              1   
29   1070935                1               1                3              1   
..       ...              ...             ...              ...            ...   
669  1350423                5              10               10              8   
670  1352848                3              10                7              8   
671  1353092                3               2                1              2   
672  1354840                2               1                1              1   
673  1354840                5               3                2              1   
674  1355260                1               1                1              1   
675  1365075                4               1                4              1   
676  1365328                1               1                2              1   
677  1368267                5               1                1              1   
678  1368273                1               1                1              1   
679  1368882                2               1                1              1   
680  1369821               10              10               10             10   
681  1371026                5              10               10             10   
682  1371920                5               1                1              1   
683   466906                1               1                1              1   
684   466906                1               1                1              1   
685   534555                1               1                1              1   
686   536708                1            

# Attribute Information:

#### 1. Sample code number: id number
#### 2. Clump Thickness: 1 - 10
#### 3. Uniformity of Cell Size: 1 - 10
#### 4. Uniformity of Cell Shape: 1 - 10
#### 5. Marginal Adhesion: 1 - 10
#### 6. Single Epithelial Cell Size: 1 - 10
#### 7. Bare Nuclei: 1 - 10
#### 8. Bland Chromatin: 1 - 10
#### 9. Normal Nucleoli: 1 - 10
#### 10. Mitoses: 1 - 10
## 11. Class: (2 for benign, 4 for malignant)

### we need to predict the benign(2) and malignant(4) stage by processing the rest of the data

In [4]:
## we cannot work with the null data so we replace the data 
## with the very large negative data
df.replace('?',-99999,inplace=True)
df.head()

id  clump_thickness  unif_cell_size  unif_cell_shape  marg_adhesion  \
0  1000025                5               1                1              1   
1  1002945                5               4                4              5   
2  1015425                3               1                1              1   
3  1016277                6               8                8              1   
4  1017023                4               1                1              3   

   single_epith_cell_size bare_nuclei  bland_chrom  norm_nucleoli  mitoses  \
0                       2           1            3              1        1   
1                       7          10            3              2        1   
2                       2           2            3              1        1   
3                       3           4            3              7        1   
4                       2           1            3              1        1   

   class  
0      2  
1      2  
2      2  
3      2  
4      2

In [5]:
##now if you keep the data in the datasets which has no 
## connection with the prediction it will reduce the accuracy
## and creae a serious problem so we need to remove it first
## in this case the id of the patient is the useless column
## because there is no relation between cancer and the patient id
##so drp the id and create a new dataframe
df.drop(['id'],1,inplace=True)
df.head()


clump_thickness  unif_cell_size  unif_cell_shape  marg_adhesion  \
0                5               1                1              1   
1                5               4                4              5   
2                3               1                1              1   
3                6               8                8              1   
4                4               1                1              3   

   single_epith_cell_size bare_nuclei  bland_chrom  norm_nucleoli  mitoses  \
0                       2           1            3              1        1   
1                       7          10            3              2        1   
2                       2           2            3              1        1   
3                       3           4            3              7        1   
4                       2           1            3              1        1   

   class  
0      2  
1      2  
2      2  
3      2  
4      2

In [6]:
## removing the output so that we can calculate the other
##by converting it to an array
X=np.array(df.drop(['class'],1))
## this is our feture data
y = np.array(df['class'])
#this is the feture the output

#### to use multiple algorithm we need a testing data sets and training data sets so that we can ind the accuracy of and test the best algorithm first we have to split it up 

In [7]:
X_train,X_test,y_train,y_test = cross_validation.train_test_split(X,y,test_size=0.2)


## test_size=.2 means we take the 20% data to test .more data needs more time 
## its enough 

In [8]:
## initiate all the algorithm package class in object

In [9]:
# before that we need to create a list in where we store accuracy


## KNN  Classifier Starts here

In [10]:
knn_clf = neighbors.KNeighborsClassifier()

#training the data with knn algorithm
knn_clf.fit(X_train,y_train)

## find the knn_accuracy based on the test result
knn_accuracy=knn_clf.score(X_test,y_test)


In [11]:
print knn_accuracy

0.9857142857142858


## Support Vector Machine Starts here

In [12]:
svm_clf = svm.SVC()

#training the data with knn algorithm
svm_clf.fit(X_train,y_train)

## find the knn_accuracy based on the test result
svm_accuracy=svm_clf.score(X_test,y_test)

In [13]:
svm_accuracy

0.9571428571428572

## GradientBoostingClassifier Starts Here

In [14]:
gb_clf = GradientBoostingClassifier()
#training the data with knn algorithm
gb_clf.fit(X_train,y_train)

## find the knn_accuracy based on the test result
gb_accuracy=gb_clf.score(X_test,y_test)

In [15]:
gb_accuracy

0.9714285714285714

## RandomForestClassifier Starts Here

In [16]:
rfc_clf = RandomForestClassifier()
#training the data with knn algorithm
rfc_clf.fit(X_train,y_train)

## find the knn_accuracy based on the test result
rfc_accuracy=gb_clf.score(X_test,y_test)

In [17]:
rfc_accuracy

0.9714285714285714

In [18]:
## assing the value in a one dimentional list
accuracy_list = [knn_accuracy,svm_accuracy,gb_accuracy,rfc_accuracy]


In [19]:
## now we create a desicion function for finding the best
def decision(knn_accuracy,svm_accuracy,gb_accuracy,rfc_accuracy):
    data=[knn_accuracy,svm_accuracy,gb_accuracy,rfc_accuracy]
    result=max(data)
    if result == knn_accuracy:
        method = "KNN Classifier"
        working_clf = knn_clf
        
    elif result==svm_accuracy:
        method='SVC Classifier'
        working_clf = svm_clf
    elif result==gb_accuracy:
        method='GB Classifier'
        working_clf = gb_clf

    else:
        method='RFC Classifier'
        working_clf = rfc_clf
    return method,working_clf
    

In [20]:
method = decision(knn_accuracy,svm_accuracy,gb_accuracy,rfc_accuracy)

In [21]:
print "best algorithm is "+str(method[0])+"\naccuracy is "+str(knn_accuracy)

best algorithm is KNN Classifier
accuracy is 0.9857142857142858


In [22]:
best_clf = method[1]
print best_clf

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=5, p=2,
           weights='uniform')


# we got our best classifier lets test a few data

## lets test the first 50 data of the datasheet

In [23]:
def string_transform(data):
    if data==2:
        value= "benign"
    if data==4:
        value="malignant"
    return value

In [24]:

data1=[[5,1,1,1,2,1,3,1,1],
[5,4,4,5,7,10,3,2,1],
[3,1,1,1,2,2,3,1,1],
[6,8,8,1,3,4,3,7,1],
[4,1,1,3,2,1,3,1,1],
[8,10,10,8,7,10,9,7,1],
[1,1,1,1,2,10,3,1,1],
[2,1,2,1,2,1,3,1,1],
[2,1,1,1,2,1,1,1,5],
[4,2,1,1,2,1,2,1,1],
[1,1,1,1,1,1,3,1,1],
[2,1,1,1,2,1,2,1,1],
[5,3,3,3,2,3,4,4,1],
[1,1,1,1,2,3,3,1,1],
[8,7,5,10,7,9,5,5,4],
[7,4,6,4,6,1,4,3,1],
[4,1,1,1,2,1,2,1,1],
[4,1,1,1,2,1,3,1,1],
[10,7,7,6,4,10,4,1,2],
[6,1,1,1,2,1,3,1,1],
[7,3,2,10,5,10,5,4,4],
[10,5,5,3,6,7,7,10,1],
[3,1,1,1,2,1,2,1,1],
[8,4,5,1,2,-99999,7,3,1],
[1,1,1,1,2,1,3,1,1],
[5,2,3,4,2,7,3,6,1],
[3,2,1,1,1,1,2,1,1],
[5,1,1,1,2,1,2,1,1],
[2,1,1,1,2,1,2,1,1],
[1,1,3,1,2,1,1,1,1],
[3,1,1,1,1,1,2,1,1],
[2,1,1,1,2,1,3,1,1],
[10,7,7,3,8,5,7,4,3],
[2,1,1,2,2,1,3,1,1],
[3,1,2,1,2,1,2,1,1],
[2,1,1,1,2,1,2,1,1],
[10,10,10,8,6,1,8,9,1],
[6,2,1,1,1,1,7,1,1],
[5,4,4,9,2,10,5,6,1],
[2,5,3,3,6,7,7,5,1],
[6,6,6,9,6,-99999,7,8,1],
[10,4,3,1,3,3,6,5,2],
[6,10,10,2,8,10,7,3,3],
[5,6,5,6,10,1,3,1,1],
[10,10,10,4,8,1,8,10,1],
[1,1,1,1,2,1,2,1,2],
[3,7,7,4,4,9,4,8,1],
[1,1,1,1,2,1,2,1,1],
[4,1,1,3,2,1,3,1,1],
[7,8,7,2,4,8,3,8,2]]
data=np.array(data1)  
data = data.reshape(len(data1),-1)
predict = best_clf.predict(data)
for item in range(len(data1)):
    print  str(item+2)+" patient in stage "+string_transform(predict[item]) +" stage"

    

2 patient in stage benign stage
3 patient in stage malignant stage
4 patient in stage benign stage
5 patient in stage malignant stage
6 patient in stage benign stage
7 patient in stage malignant stage
8 patient in stage benign stage
9 patient in stage benign stage
10 patient in stage benign stage
11 patient in stage benign stage
12 patient in stage benign stage
13 patient in stage benign stage
14 patient in stage malignant stage
15 patient in stage benign stage
16 patient in stage malignant stage
17 patient in stage benign stage
18 patient in stage benign stage
19 patient in stage benign stage
20 patient in stage malignant stage
21 patient in stage benign stage
22 patient in stage malignant stage
23 patient in stage malignant stage
24 patient in stage benign stage
25 patient in stage benign stage
26 patient in stage benign stage
27 patient in stage malignant stage
28 patient in stage benign stage
29 patient in stage benign stage
30 patient in stage benign stage
31 patient in stage beni

# more than 97% matched exact result showed in the datasets
# thats a significant improvement
